File: sdks/java/src/main/java/com/blok/nanoservice/nodes/ApiCallNode.java
package com.blok.nanoservice.nodes;

import com.blok.nanoservice.errors.NodeException;
import com.blok.nanoservice.node.NodeHandler;
import com.blok.nanoservice.types.Context;
import com.google.gson.Gson;
import com.google.gson.JsonSyntaxException;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.HashMap;
import java.util.Map;

/**
 * Node that makes HTTP requests to external APIs.
 * <p>
 * Config:
 * <ul>
 *   <li>url (string, required) - The URL to call</li>
 *   <li>method (string, optional) - HTTP method (default: "GET")</li>
 *   <li>timeout (number, optional) - Timeout in seconds (default: 10)</li>
 *   <li>headers (map, optional) - Additional request headers</li>
 * </ul>
 * Input (request body):
 * <ul>
 *   <li>body (any, optional) - Request body for POST/PUT/PATCH</li>
 * </ul>
 * Output:
 * <ul>
 *   <li>status (number) - HTTP status code</li>
 *   <li>data (any) - Parsed response body</li>
 *   <li>headers (map) - Response headers</li>
 * </ul>
 */
public class ApiCallNode implements NodeHandler {

    private static final Gson GSON = new Gson();

    @Override
    public Object execute(Context ctx, Map<String, Object> config) throws Exception {
        if (config == null) {
            throw NodeException.configuration("config is required for api-call node");
        }

        // Get URL from config (required)
        Object urlVal = config.get("url");
        if (!(urlVal instanceof String url) || url.isEmpty()) {
            throw NodeException.configuration("'url' is required in node config");
        }

        // Get method from config, default to GET
        String method = "GET";
        Object methodVal = config.get("method");
        if (methodVal instanceof String m && !m.isEmpty()) {
            method = m.toUpperCase();
        }

        // Get timeout from config, default to 10 seconds
        int timeout = 10;
        Object timeoutVal = config.get("timeout");
        if (timeoutVal instanceof Number n && n.intValue() > 0) {
            timeout = n.intValue();
        }

        // Build the HTTP request
        HttpRequest.Builder requestBuilder = HttpRequest.newBuilder()
                .uri(URI.create(url))
                .timeout(Duration.ofSeconds(timeout));

        // Set custom headers from config
        Object headersVal = config.get("headers");
        if (headersVal instanceof Map<?, ?> headersMap) {
            for (Map.Entry<?, ?> entry : headersMap.entrySet()) {
                if (entry.getKey() instanceof String key && entry.getValue() instanceof String value) {
                    requestBuilder.header(key, value);
                }
            }
        }

        // Prepare request body
        String requestBody = null;
        Map<String, Object> body = ctx.getRequest().bodyMap();
        if (body != null && body.containsKey("body") && body.get("body") != null) {
            requestBody = GSON.toJson(body.get("body"));
        }

        // Set method and body
        switch (method) {
            case "POST" -> {
                requestBuilder.header("Content-Type", "application/json");
                requestBuilder.POST(HttpRequest.BodyPublishers.ofString(
                        requestBody != null ? requestBody : ""));
            }
            case "PUT" -> {
                requestBuilder.header("Content-Type", "application/json");
                requestBuilder.PUT(HttpRequest.BodyPublishers.ofString(
                        requestBody != null ? requestBody : ""));
            }
            case "DELETE" -> requestBuilder.DELETE();
            default -> requestBuilder.GET();
        }

        // Execute the request
        try {
            HttpClient client = HttpClient.newBuilder()
                    .connectTimeout(Duration.ofSeconds(timeout))
                    .build();

            HttpResponse<String> response = client.send(
                    requestBuilder.build(),
                    HttpResponse.BodyHandlers.ofString());

            // Try to parse response as JSON, fall back to string
            Object responseData;
            try {
                responseData = GSON.fromJson(response.body(), Object.class);
            } catch (JsonSyntaxException e) {
                responseData = response.body();
            }

            // Collect response headers
            Map<String, String> responseHeaders = new HashMap<>();
            response.headers().map().forEach((key, values) -> {
                if (!values.isEmpty()) {
                    responseHeaders.put(key, values.get(0));
                }
            });

            Map<String, Object> result = new HashMap<>();
            result.put("status", response.statusCode());
            result.put("data", responseData);
            result.put("headers", responseHeaders);
            return result;

        } catch (Exception e) {
            if (e instanceof NodeException) {
                throw e;
            }
            throw NodeException.network("request to " + url + " failed: " + e.getMessage(), e);
        }
    }
}

File: sdks/php/examples/ApiCallNode.php
<?php

declare(strict_types=1);

namespace Blok\Blok\Examples;

use Blok\Blok\Errors\NodeException;
use Blok\Blok\Node\NodeHandler;
use Blok\Blok\Types\Context;

/**
 * ApiCallNode makes HTTP requests to external APIs.
 *
 * Config:
 *   - url (string, required): The URL to call
 *   - method (string, optional): HTTP method (default: GET)
 *   - timeout (int, optional): Timeout in seconds (default: 10)
 *   - headers (object, optional): Additional headers to send
 *
 * Body (for POST/PUT/PATCH):
 *   - body (mixed, optional): Request body to send
 */
final class ApiCallNode implements NodeHandler
{
    public function execute(Context $ctx, array $config): mixed
    {
        $url = $config['url'] ?? null;
        if (!is_string($url) || $url === '') {
            throw NodeException::configuration("'url' is required in node config");
        }

        $method = strtoupper((string) ($config['method'] ?? 'GET'));
        $timeout = (int) ($config['timeout'] ?? 10);

        // Build HTTP context options
        $httpOptions = [
            'method' => $method,
            'timeout' => $timeout,
            'ignore_errors' => true,
            'header' => "Content-Type: application/json\r\n",
        ];

        // Add custom headers from config
        if (isset($config['headers']) && is_array($config['headers'])) {
            foreach ($config['headers'] as $key => $value) {
                if (is_string($value)) {
                    $httpOptions['header'] .= sprintf("%s: %s\r\n", $key, $value);
                }
            }
        }

        // Add body for POST/PUT/PATCH
        if (in_array($method, ['POST', 'PUT', 'PATCH'], true)) {
            $body = is_array($ctx->request->body) ? ($ctx->request->body['body'] ?? null) : null;
            if ($body !== null) {
                $httpOptions['content'] = json_encode($body, JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE);
            }
        }

        $streamContext = stream_context_create([
            'http' => $httpOptions,
            'ssl' => [
                'verify_peer' => true,
                'verify_peer_name' => true,
            ],
        ]);

        $responseBody = @file_get_contents($url, false, $streamContext);

        if ($responseBody === false) {
            throw NodeException::network(sprintf('request to %s failed', $url));
        }

        // Parse status code from response headers
        $statusCode = 200;
        $responseHeaders = [];
        if (isset($http_response_header) && is_array($http_response_header)) {
            foreach ($http_response_header as $header) {
                if (preg_match('/^HTTP\/[\d.]+ (\d{3})/', $header, $matches)) {
                    $statusCode = (int) $matches[1];
                } else {
                    $parts = explode(':', $header, 2);
                    if (count($parts) === 2) {
                        $responseHeaders[strtolower(trim($parts[0]))] = trim($parts[1]);
                    }
                }
            }
        }

        // Try to parse response as JSON, fall back to string
        $data = json_decode($responseBody, true);
        if (json_last_error() !== JSON_ERROR_NONE) {
            $data = $responseBody;
        }

        return [
            'status' => $statusCode,
            'data' => $data,
            'headers' => $responseHeaders,
        ];
    }
}
